src/ui/public/visualize/loader/embedded_visualize_handler.js

```diff
@@ -26,58 +30,144 @@ const RENDER_COMPLETE_EVENT = 'render_complete';
  * with the visualization.
  */
 export class EmbeddedVisualizeHandler {
-  constructor(element, scope, savedObject) {
+  constructor(element, savedObject, params) {
     this._element = element;
-    this._scope = scope;
     this._savedObject = savedObject;
+    this._params = {
+      uiState: params.uiState,
+      queryFilter: params.queryFilter,
+      searchSource: savedObject.searchSource,
+      aggs: savedObject.vis.getAggConfig(),
+      timeRange: params.timeRange,
+      filters: params.filters,
+      query: params.query,
+    };
+
     this._listeners = new EventEmitter();
     // Listen to the first RENDER_COMPLETE_EVENT to resolve this promise
     this._firstRenderComplete = new Promise(resolve => {
       this._listeners.once(RENDER_COMPLETE_EVENT, resolve);
     });
-    this._element.on('renderComplete', () => {
+    this._element.addEventListener('renderComplete', () => {
       this._listeners.emit(RENDER_COMPLETE_EVENT);
     });
+
+    this._init(params);
   }
 
+  _init(params) {
+    this._loaded = false;
+    this._destroyed = false;
+
+    this._appState = params.appState;
+    this._vis = this._savedObject.vis;
+    this._vis._setUiState(params.uiState);
+    this._uiState = this._vis.getUiState();
+
+    this._vis.on('update', this._handleVisUpdate);
+    this._vis.on('reload', this._reloadVis);
+    this._uiState.on('change', this._fetchAndRender);
+
+    this._visualize = new VisualizeDataLoader(this._vis, params.Private);
+    this._renderCompleteHelper = new RenderCompleteHelper(this._element);
+
+    this._render();
+  }
+
+  _updateState = (visState) => {
```

Comment:
I don't feel strongly about this, but personally I'd find this a bit clearer if this logic was moved into `_handleVisUpdate`. Because it's extracted I wrongly assume that it's called from multiple places, but it's only called by the handler.